vispy/visuals/visual.py

```diff
@@ -266,6 +267,28 @@ def update(self):
     def _transform_changed(self, event=None):
         self.update()
 
+    def push_gl_state(self, *args, **kwargs):
+        raise NotImplementedError(self)
+
+    def pop_gl_state(self):
+        raise NotImplementedError(self)
+
+    def set_gl_state(self, *args, **kwargs):
+        raise NotImplementedError(self)
+
+    def update_gl_state(self, *args, **kwargs):
+        raise NotImplementedError(self)
+
+    @contextmanager
+    def temp_gl_state(self, *args, **kwargs):
```

Comment:
Any better names for this that aren't already used? And why does this need to be in the BaseVisual where there is no concept of GL state (right? That's all in the `Visual`?).

It's been a while since I've tried to do something like this myself, but can we have `set_gl_state` return an object that defines an `__enter__/__exit__` and on `__enter__` does nothing (because the state is already set), but on `__exit__` it resets/pops the state back to the original value? Then `set_gl_state` is the function you can call and the context manager.

Reply:
Thanks - that sounds great. I'll figure that out if it's possible! I didn't like this name either.

Reply:
Don't get distracted by the complexity of what the `__exit__` method is doing, but here is what `donfig` does (which is based on dask's config functionality):

https://github.com/pytroll/donfig/blob/f75ee608e9e57225084b9a4d06ba449010308bb0/donfig/config_obj.py#L320-L337

This is part of the `ConfigSet` object which is what is returned by the main `Config.set` method. So the `Config` object isn't a context manager, but the object `.set` returns is.

Reply:
Thanks for taking time to look at this and for all the suggestions and pointers. I _think_ this working as way you described and it feels much better to me.

Please check out the tests in `vispy/scene/tests/test_visuals.py` to get an idea of the functionality now. There are likely more improvements to be made, so I of course welcome any more feedback!